plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/concurrent/LSPExecutorService.java

```diff
@@ -0,0 +1,296 @@
+/**
+ * Copyright (c) 2020 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.xtext.server.concurrent;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.log4j.Logger;
+import org.eclipse.xtext.service.OperationCanceledManager;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * Similar to an ordinary {@link ExecutorService}, but with two differences:
+ * <ol>
+ * <li>when submitting a task, a queue ID may be provided and this executor will then guarantee that
+ * <ol type="a">
+ * <li>two tasks with {@link Object#equals(Object) equal} queue ID are never running at the same time, and
+ * <li>tasks with equal queue ID will be executed in the order they were submitted.
+ * </ol>
+ * If no queue ID is provided, the usual behavior of ordinary {@link ExecutorService executor services} applies, i.e.
+ * the task might run concurrently to any other task and ordering depends on the {@link #delegate delegate executor
+ * service}.
+ * <li>each task is provided with an Xtext {@link CancelIndicator} and cancellation via a task's
+ * {@link CompletableFuture future}, as returned by the submit methods, is converted into a cancellation request via
+ * that cancel indicator. This leads to a slight change how cancellation works compared to ordinary futures, see
+ * {@link QueuedTaskFuture#cancel(boolean) #cancel(boolean)} for details.
+ * </ol>
+ *
+ * <h2>Cancellation Handling</h2>
+ *
+ * Cancellation is handled entirely through Xtext's {@link CancelIndicator}s and a task's implementation can decide how
+ * to react to that (as usual with cancel indicators). Tasks will always be invoked, even if they were marked as
+ * cancelled while waiting on the queue, so implementors of a task may choose to implement a "non-cancellable" operation
+ * at the beginning of a task before reacting to the cancel indicator.
+ */
+// FIXME remove @Singleton?
+@Singleton // ensures queue IDs are "global" within a single injector
+public class LSPExecutorService {
```

Comment:
This design looks like a simplified equivalent to Eclipses SchedulingRule.
What I don't understand is where a guarantee comes from that two tasks with different descriptions don't interfere with each other. As far as I can see, the description is usually the URI of the open resource of an editor (an editor id). This will sequentialise requests per editor. So far so good.

Now if two circular dependent editors run tasks that will be re-entrant on the executor service, we will dead-lock. How can we guarantee that this is not going to happen?

Reply:
`LSPExectorService` (with its queueId approach) and the `OpenFilesManager#runInOpen|TemporaryFileContext()` methods are not intended for fine-grained operations/computations that must be composable (see also my answer to your previous question). They correspond more to `XtextDocument#readOnly|modify()` when invoked on an XtextEditor's document, than to Eclipse's scheduling rules.

Another way to put it: the scenario you described ("Now if two circular ...") could also be replicated in the context of Eclipse if one Eclipse Xtext editior A would run an operation in another Xtext editor B via `getDocument().readOnly|modify()` which would then try to obtain a lock back on A. But this is unlikely because `XtextDocument#readOnly|modify()` is not intended for such composable operations. In 99% of cases editor should stick to their own context or create temporary, entirely unrelated and self-contained contexts.

However, it would probably be possible to detect such scenarios in `OpenFilesManager#runInOpen|TemporaryFileContext()` (or even in `LSPExecutorService`?) to at least fail more or less gracefully instead of running into a dead lock.

If we find that we need `OpenFilesManager#runInOpen|TemporaryFileContext()` more often and for more cases than I had in mind, we might have to introduce a separation of read-only and read/write operations in OpenFilesManager or some other locking mechanism, but I wanted to avoid this as long as we don't run into a true need for it.

Reply:
> the scenario you described ("Now if two circular ...") could also be replicated in the context of Eclipse

In theory yes, practically all readOnlys of that shape in Xtext Eclipse would be implicitly sequential since they are triggered from the UI thread.

Reply:
So you mean in Eclipse editors do not actually run in parallel?

If I make an edit in one of 10 open editors (not saving) and all other editors are affected by that edit, then scoping, resolution, validation, etc. of the other 9 editors is done one after another?

If that is the case and it never caused lag in Eclipse, we could think about that for LSP, too. So far, my goal on this branch was to allow to distribute work for open editors across multiple threads.

Reply:
Hmm that's not what I meant.
In Eclipse, we have the UI thread. All user interaction is triggered from there. When a document is changed, a document event is scheduled and after a delay, the resource will be refreshed while holding a modify lock on the document. Linking etc will start while holding that lock and afterwards the other editors will receive a notification. They will be processed in own jobs subsequently, each holding a lock per document.

What I was referring to: This logic is under almost full control of the framework. That's why we "know" that the code won't escape to another document and attempt to do a modify().

Now back to the API on IXtextDocument - readOnly and modify. If a client implements code that aquires more than one modify on different docs, it could potentially deadlock, too. However, the modify(..) is usually only triggered from code that runs on the UI thread. Therefore it's would require other background threads that try to do concurrent modifies. In that case, we'd deadlock.

But: since that's highly unlikey, we won't see that. With LSP and Commands, that are executed on their own thread, the sequential UI thread is not present.

Reply:
That's more or less what I had in mind and wanted to achieve in the LSP case too: that all the tricky stuff is handled by the framework. The difference of not starting out on the UI thread is a good point, though. Let's discuss this further once you are through with the diff.

Reply:
After thinking about this some more, one additional comment about the deadlock scenario:

> Now if two circular dependent editors run tasks that will be re-entrant on the executor service, we will dead-lock. How can we guarantee that this is not going to happen?

With `OpenFilesManager`/`LSPExecutorService` a deadlock can occur in the scenario you mention only if one of the tasks being submitted/executed via `#runInOpenFileContext()` synchronously waits for the result of another task it (transitively) submitted. But this is easy to avoid because code that requires the result can be put into a completion stage (i.e. with `#thenApply()` etc. on the completable future returned by `#runInOpenFileContext()` when submitting the transitive task).

The only time we have to wait for a result is when existing API forces us to do so. So far, I have only found one case of this: `#readOnly(URI, IUnitOfWork<R, ResourceSet>)` in `IReferenceFinder.IResourceAccess` which is, unfortunately, used in LSP for more purposes than just reference finding. Hopefully we can refactor this or make sure it won't be a problem in other ways.

As a side note: this situation seems to be different with `XtextDocument` and Eclipse editors, because `XtextDocument#readOnly|modify()` return the result itself, not a future.

Having said all this, my comment above that `#runInOpenFileContext()` is not intended for use with fine-grained, composable operations still applies. When in the context of open file A the resource of another file B is required, that resource can simply be loaded into the resource set of A's open file context without switching contexts (at least in the standard case and once an IExternalContentSupport is in place).

Reply:
> i.e. with #thenApply()

Wouldn't this escape the safety net that the OpenFilesManager tries to implement, e.g. only a single task with a given ID may run?
If I submit such a task and do a thenApply on the future, my previously isolated environment is no longer available. Or do I miss something?

Anyhow, if this is not the desired use-case, my feedback is more of academic value than practically relevant.

Reply:
> ... my previously isolated environment is no longer available.

True! If access to the resource/document is required in the later stage as well, another `#runInOpenFileContext()` is needed. It's a bit nasty, but the same (or very similar) situation exists for the implementation of the transitive task being submitted to run in context of resource B: that must not access resource A either, even though it will often have access to the object reference in the code when implemented as a closure.

> Anyhow, if this is not the desired use-case, my feedback is more of academic value than practically relevant.

In any case, this discussion was very helpful for me! I will try to extract 2-3 recommendations and add them to the API doc somewhere.